### PR TITLE
Add model checkpointing to push_to_hub and PushToHubCallback

### PR DESCRIPTION
**Pull request by Rocketknight1**

### CODE REVIEW

File: src/transformers/file_utils.py [hunk 1]
```diff
@@ -2284,6 +2295,19 @@ def push_to_hub(
...
 
+    def save_checkpoint(self, checkpoint_dir, epoch):
+        if not os.path.isdir(checkpoint_dir):
+            os.mkdir(checkpoint_dir)
+        # We avoid tf.train.checkpoint or saving weights in TF format, even though that includes optimizer
+        # state for us, because it requires special handling for objects like custom losses, which we use
+        # internally and which users are likely to use too
+        weights_path = os.path.join(checkpoint_dir, "weights.h5")
+        self.save_weights(weights_path)
+        extra_data = {"epoch": epoch, "optimizer_state": self.optimizer.get_weights()}
```
Comment (sgugger) [inline src/transformers/file_utils.py:2306]: Don't we need the learning rate scheduler state as well?

**Review by sgugger — APPROVED** [review]
Thanks for adding this to the callback!

File: src/transformers/file_utils.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/file_utils.py:2306]: Good point, I'll add that! I'm a bit worried that we have to manually list the things we're saving - `tf.train.Checkpoint` is supposed to take care of this, but the fact that it doesn't co-operate with custom losses has been a constant headache.

Comment (Rocketknight1) [reply src/transformers/file_utils.py:2306]: Actually, wait, no need - `save_weights` also includes the iteration count!

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -753,6 +754,39 @@ def get_input_embeddings(self) -> tf.keras.layers.Layer:
...
+        with open(extra_data_path, "wb") as f:
+            pickle.dump(extra_data, f)
+
+    def load_repo_checkpoint(self, repo_path_or_name, organization=None):
+        if getattr(self, "optimizer", None) is None:
+            raise RuntimeError(
+                "Checkpoint loading failed as no optimizer is attached to the model. "
+                "This is most likely caused by the model not being compiled."
+            )
+        repo = self._create_or_get_repo(repo_path_or_name, organization)
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:776]: I think we said to use `Repository` here to avoid creating a new repo?

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -753,6 +754,39 @@ def get_input_embeddings(self) -> tf.keras.layers.Layer:
...
+        weights_file = os.path.join(checkpoint_dir, "weights.h5")
+        if not os.path.isfile(weights_file):
+            raise FileNotFoundError(f"Could not find checkpoint file weights.h5 in repo {repo_path_or_name}!")
+        extra_data_file = os.path.join(checkpoint_dir, "extra_data.pickle")
+        if not os.path.isfile(extra_data_file):
+            raise FileNotFoundError(f"Could not find checkpoint file extra_data.pickle in repo {repo_path_or_name}!")
+        self.load_weights(weights_file)
+        with open(extra_data_file, "rb") as f:
+            extra_data = pickle.load(f)
+        self.optimizer.set_weights(extra_data["optimizer_state"])
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:787]: This could use new lines to separate blocks and comments on what each file does for the non-expert Keras users :-)

File: src/transformers/keras_callbacks.py [hunk 4]
```diff
@@ -48,8 +50,13 @@ def __init__(
         hub_token (:obj:`str`, `optional`):
             The token to use to push the model to the Hub. Will default to the token in the cache folder obtained with
             :obj:`huggingface-cli login`.
+        checkpoint (:obj:`bool`, `optional`):
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:53]:
```suggestion
        checkpoint (:obj:`bool`, `optional`, defaults to :obj:`False`):
```

File: src/transformers/keras_callbacks.py [hunk 5]
```diff
@@ -23,6 +24,7 @@ def __init__(
         tokenizer: Optional[PreTrainedTokenizerBase] = None,
         hub_model_id: Optional[str] = None,
         hub_token: Optional[str] = None,
+        checkpoint: Optional[bool] = False,
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:27]:
```suggestion
        checkpoint: bool = False,
```
It's only `Optional` if it can be `None`.

**Review by sgugger — APPROVED** [review]
Added a few last comments!

File: src/transformers/modeling_tf_utils.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Done!

File: src/transformers/modeling_tf_utils.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:791]: I tried this, but it resulted in some code duplication in order to make everything work - I kept it as-is but added a comment to explain instead. If you'd still prefer to avoid the call, let me know!

Comment (sgugger) [reply src/transformers/modeling_tf_utils.py:791]: Let's see what @LysandreJik thinks then.

Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py:791]: Are you sure it won't actually create a remote repository if it doesn't exist? If you're certain of it it's fine for me to go with `self._create_or_get_repo` but I'm quite unsure that it won't create a remote repo.

Also which method are you using this method here?

https://github.com/huggingface/transformers/blob/956a483173e77ebf655ca9636a5f7b6ef010b307/src/transformers/file_utils.py#L2338-L2367

Because if so, don't call it with positional arguments like you do here as you're passing the `organization` parameter to `repo_url`.

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:791]: Good spot on the `organization` positional arg, fixed!

Also, the `_create_or_get_repo` method just does this, after sanity-checking the arguments:

```
repo = Repository(repo_path_or_name, clone_from=repo_url, use_auth_token=use_auth_token)
repo.git_pull()
return repo
```
So I think as long as we don't `push` it afterwards, we won't create a new remote repo if one doesn't exist already.

Comment (sgugger) [reply src/transformers/modeling_tf_utils.py]: I think this discussion is real proof we need to just put the code used here and not try to use some magic of the Mixin. The library always prefers an explicit approach at the cost of duplicate code, and I think this is an instance where we should just apply the code needed.

The Mixin might also change in the feature, following the RFC on huggingface_hub, so those private methods might actually introduce some behavior we don't want in the future.

Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py]: And the `Repository(repo_path_or_name, clone_from=repo_url, use_auth_token=use_auth_token)` line actually creates the repo if you try to clone it but it doesn't exist:

https://github.com/huggingface/huggingface_hub/blob/10c69146969ad7f9e1add075c1ef4ec15e42e85f/src/huggingface_hub/repository.py#L542-L549

Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py]: The `push` event is unrelated to the creation of the repository. No push to a remote repository can be done if the repository does not exist. The repository creation happens at the repository creation, not at the repository update (push).

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Ah, I'm sorry, I didn't realize that's how it worked! And yes, you're right, this is probably a sign that we need to be explicit. I'll try to figure out some code that just checks without accidentally creating a new repo.

File: src/transformers/modeling_tf_utils.py [hunk 6]
```diff
@@ -753,6 +756,69 @@ def get_input_embeddings(self) -> tf.keras.layers.Layer:
...
+                "This is most likely caused by the model not being compiled."
+            )
+        if not os.path.isdir(repo_path_or_name):
+            # If this isn't a local path, check that the remote repo exists and has a checkpoint in it
+            repo_files = list_repo_files(repo_path_or_name)
+            for file in ("checkpoint/weights.h5", "checkpoint/extra_data.pickle"):
+                if file not in repo_files:
+                    raise FileNotFoundError(f"Repo {repo_path_or_name} does not contain checkpoint file {file}!")
+            # This function is safe now that we know the repo exists
+            repo = self._create_or_get_repo(repo_path_or_name)
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:797]: Can we just stop using this private function and create the `Repository` object?
As I mentioned before this function will probably be removed entirely or reworked in a different way.

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Done!

File: src/transformers/modeling_tf_utils.py [hunk 7]
```diff
@@ -753,6 +756,74 @@ def get_input_embeddings(self) -> tf.keras.layers.Layer:
...
+                "Checkpoint loading failed as no optimizer is attached to the model. "
+                "This is most likely caused by the model not being compiled."
+            )
+        if not os.path.isdir(repo_path_or_name):
+            # If this isn't a local path, check that the remote repo exists and has a checkpoint in it
+            repo_files = list_repo_files(repo_path_or_name)
+            for file in ("checkpoint/weights.h5", "checkpoint/extra_data.pickle"):
+                if file not in repo_files:
+                    raise FileNotFoundError(f"Repo {repo_path_or_name} does not contain checkpoint file {file}!")
+            # This function is safe now that we know the repo exists
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:796]: Not necessary anymore.
```suggestion
```

**Review by sgugger — APPROVED** [review]
Thanks for adjusting!
